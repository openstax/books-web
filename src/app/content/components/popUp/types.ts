import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
