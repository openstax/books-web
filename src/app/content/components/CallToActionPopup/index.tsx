import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import { loginLink as loginLinkSelector, signupLink as signupLinkSelector } from '../../../auth/selectors';
import Button from '../../../components/Button';
import htmlMessage from '../../../components/htmlMessage';
import { closeCallToActionPopup } from '../../actions';
import { showCTAPopup } from '../../selectors';
import {
  CTABottomLayer,
  CTAButtons,
  CTACloseButton,
  CTACloseIcon,
  CTAContent,
  CTAGraphic,
  CTAHeading,
  CTAMiddleLayer,
  CTAParagraph,
  CTAText,
  CTATextLink,
  CTATopLayer,
  CTAWrapper,
} from './styles';

// tslint:disable-next-line: variable-name
const CallToActionPopup = () => {
  const dispatch = useDispatch();
  const showPopup = useSelector(showCTAPopup);
  const loginLink = useSelector(loginLinkSelector);
  const signupLink = useSelector(signupLinkSelector);

  if (!showPopup) { return null; }

  const closePopup = () => {
    dispatch(closeCallToActionPopup());
  };

  return <CTAWrapper data-analytics-region='signup CTA'>
    <CTABottomLayer />
    <CTAMiddleLayer />
    <CTATopLayer />
    <CTAContent>
      <CTAText>
        <CTAHeading>
          <FormattedMessage id='i18n:cta:heading'>
            {(msg: Element | string) => msg}
          </FormattedMessage>
        </CTAHeading>
        <CtaPara />
        <CTAButtons>
          <FormattedMessage id='i18n:cta:button:signup'>
            {(msg: Element | string) => <Button
              data-testid='sign-up'
              data-analytics-label='sign-up'
              size='large'
              variant='primary'
              // eslint-disable-next-line
              component={<a href={signupLink}/>}
            >{msg}</Button>}
          </FormattedMessage>
          <CTATextLink>
            <FormattedMessage id='i18n:cta:button:separator'>
              {(msg: Element | string) => msg}
            </FormattedMessage>
            <FormattedMessage id='i18n:cta:button:login'>
              {(msg: Element | string) => <a
                data-testid='login'
                data-analytics-label='login'
                href={loginLink}
              >{msg}</a>}
            </FormattedMessage>
          </CTATextLink>
        </CTAButtons>
      </CTAText>
      <CTAGraphic />
    </CTAContent>
    <CTACloseButton onClick={closePopup} data-analytics-label='close' data-testid='closeCTA'>
      <CTACloseIcon />
    </CTACloseButton>
  </CTAWrapper>;
};

export default CallToActionPopup;

// tslint:disable-next-line: variable-name
const CtaPara = htmlMessage('i18n:cta:text', CTAParagraph);
