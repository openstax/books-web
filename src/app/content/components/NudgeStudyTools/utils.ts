import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides, studyGuidesEnabled } from '../../studyGuides/selectors';
import {
  arrowDesktopHeight,
  arrowLeftMargin,
  arrowMobileHeight,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  contentMarginTop,
  cookieNudge,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  spotlightPadding,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperRight: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean, windowWidth: number): Positions => {
  const { top, left, right, height, width } = getBoundingRectOfNudgeTarget(target);
  const padding = remsToPx(spotlightPadding);
  const spotlightTopOffset = top - padding;
  const spotlightLeftOffset = left - padding;
  const spotlightHeight = height + (padding * 2);
  const spotlightWidth = width + (padding * 2);
  const arrowTopOffset = spotlightTopOffset + spotlightHeight + remsToPx(arrowTopMargin);
  const arrowLeft = spotlightLeftOffset + remsToPx(arrowLeftMargin);
  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(isMobile ? arrowMobileHeight : arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperRight = windowWidth - right - padding;
  const closeButtonLeft = left + spotlightWidth;
  const closeButtonTopOffset = contentWrapperTopOffset - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperRight,
    contentWrapperTopOffset,
    spotlightHeight,
    spotlightLeftOffset,
    spotlightTopOffset,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  // Starting values depends on if we checks for Math.max or Math.min
  let top = 9999;
  let bottom = 0;
  let left = 9999;
  let right = 0;

  for (const child of Array.from(target.children)) {
    const rect = child.getBoundingClientRect();
    top = Math.min(top, rect.top);
    bottom = Math.max(bottom, rect.bottom);
    left = Math.min(left, rect.left);
    right = Math.max(right, rect.right);
  }

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);
  const isEnabled = useSelector(studyGuidesEnabled);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(`#${nudgeStudyToolsTargetId}`) as HTMLElement | null;
    if (isEnabled && target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results with different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile, windowWidth));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [studyGuides, isEnabled, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  Cookies.set(cookieNudge.pageCounter, (counter + 1).toString());
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString());
  Cookies.set(cookieNudge.date, now.toString());
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  return React.useEffect(() => {
    const counter = getPageCounterCookie();
    if (page && counter <= nudgeStudyToolsMinPageLimit) { incrementPageCounterCookie(); }
  }, [page]);
};
