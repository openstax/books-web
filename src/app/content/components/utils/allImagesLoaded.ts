import { Document,  Element } from '@openstax/types/lib.dom';

export default (container: Element | Document) => {
  return Promise.all(Array.from(container.querySelectorAll('img')).map((img) => img.complete
    ? Promise.resolve()
    : new Promise<void>((resolve) => {

      const previousOnload = img.onload;
      const previousOnerror = img.onerror;

      img.onload = (...args) => {
        resolve();
        if (previousOnload) {
          previousOnload.apply(img, args);
        }
      };
      img.onerror = (...args) => {
        resolve();
        if (previousOnerror) {
          previousOnerror.apply(img, args);
        }
      };
    })
  )).then((): void => undefined);
};
