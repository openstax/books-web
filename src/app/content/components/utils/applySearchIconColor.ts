import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { SearchButtonColor } from '../../types';
import { toolbarIconColor } from '../constants';

export const applySearchIconColor = (props: {colorSchema: SearchButtonColor}) =>
  props.colorSchema ? css`
    color: ${theme.color.primary[props.colorSchema].foreground};

    :hover,
    :focus {
        color: ${theme.color.primary[props.colorSchema].foreground};

        ::before {
            content:"";
            display: block;
            height: 100%;
            position: absolute;
            top: 0;
            right: 0;
            min-width: 45px;
            background-color: rgba(0, 0, 0, 0.06)
            ;
        }
    }
  ` : css`
  color: ${toolbarIconColor.base};
`;
