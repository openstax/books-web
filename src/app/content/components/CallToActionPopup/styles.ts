import styled from 'styled-components';
import { Times } from 'styled-icons/fa-solid/Times/Times';
import Button from '../../../components/Button';
import { textRegularStyle } from '../../../components/Typography';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';

// tslint:disable-next-line: variable-name
export const CTAWrapper = styled.div`
  position: fixed;
  left: 0;
  bottom: 0;
  z-index: 99;
  width: 100%;

  @media screen and (max-width: 1200px) {
    display: none;
  }
`;

// tslint:disable-next-line: variable-name
export const CTAContent = styled.div`
  position: absolute;
  z-index: 3;
  bottom: 0;
  width: 100%;
  display flex;
  justify-content: center;
  align-items: center;
`;

// tslint:disable-next-line: variable-name
export const CTAText = styled.div``;

// tslint:disable-next-line: variable-name
export const CTAHeading = styled.h2`
  ${textStyle}
  font-size: 3.6rem;
  line-height: 4rem;
  color: ${theme.color.text.white};
  margin: 0;
  max-width: 570px;
  overflow: hidden;
`;

// tslint:disable-next-line: variable-name
export const CTAParagraph = styled.p`
  ${textStyle}
  font-size: 1.8rem;
  line-height: 3rem;
  margin: 1rem 0;
  color: ${theme.color.text.white};
  max-width: 570px;
`;

// tslint:disable-next-line: variable-name
export const CTAButtons = styled.div`
  display: flex;
  align-items: center;
  ${Button} {
    width: 170px;
  }
`;

// tslint:disable-next-line: variable-name
export const CTATextLink = styled.span`
  ${textRegularStyle}
  color: ${theme.color.text.white};
  margin-left: 1.5rem;
  a {
    color: ${theme.color.text.white};
    text-decoration: underline;
    margin-left: 0.5rem;
  }
`;

// tslint:disable-next-line: variable-name
export const CTACloseButton = styled(Times)`
  position: absolute;
  top: 40%;
  right: 3rem;
  z-index: 3;
  width: 1.8rem;
  color: ${theme.color.text.white};
  cursor: pointer;
`;

// tslint:disable-next-line: variable-name
export const CTAGraphic = styled.img`
  margin-left: 3rem;
`;

// tslint:disable-next-line: variable-name
export const CTATopLayer = styled.img`
  position: absolute;
  right: -5px;
  bottom: 0;
  z-index: 2;
`;

// tslint:disable-next-line: variable-name
export const CTAMiddleLayer = styled.img`
  width: 100%;
  position: absolute;
  left: 0;
  bottom: 0;
  z-index: 1;
`;

// tslint:disable-next-line: variable-name
export const CTABottomLayer = styled.img`
  width: 100%;
  position: absolute;
  left: 0;
  bottom: 20px;
  z-index: 0;
`;

// tslint:disable-next-line: variable-name
export const CTABackground = styled.div`
  position: relative;
  height: 420px;
  overflow: hidden;
`;
