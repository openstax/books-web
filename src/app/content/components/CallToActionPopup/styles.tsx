import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import { Times } from 'styled-icons/fa-solid/Times/Times';
import Button from '../../../components/Button';
import { textRegularStyle } from '../../../components/Typography';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import bottomLayer from './assets/bottom-layer.svg';
import ctaGraphic from './assets/desktop-mobile-graphic.svg';
import middleLayer from './assets/middle-layer.svg';
import topLayer from './assets/top-layer.svg';

const slideIn = keyframes`
  0% {
    bottom: -100%;
  }

  100% {
    bottom: 0;
  }
`;

const slideInAnimation = css`
  animation: ${300}ms ${slideIn} ease-out;
`;

// tslint:disable-next-line: variable-name
export const CTAWrapper = styled.div`
  position: fixed;
  left: 0;
  bottom: 0;
  z-index: ${theme.zIndex.ctaPopup};
  width: 100%;
  ${slideInAnimation}
  ${theme.breakpoints.mobile(css`
    display: none;
  `)}
`;

// tslint:disable-next-line: variable-name
export const CTAContent = styled.div`
  position: absolute;
  bottom: 0;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1;
`;

// tslint:disable-next-line: variable-name
export const CTAHeading = styled.h2`
  ${textStyle}
  font-size: 3.6rem;
  line-height: 4rem;
  letter-spacing: -0.14rem;
  color: ${theme.color.text.white};
  margin: 0;
  max-width: 50rem;
  overflow: hidden;
`;

// tslint:disable-next-line: variable-name
export const CTAParagraph = styled.p`
  ${textStyle}
  font-size: 1.8rem;
  line-height: 3rem;
  margin: 1rem 0;
  color: ${theme.color.text.white};
  max-width: 57rem;
`;

// tslint:disable-next-line: variable-name
export const CTAButtons = styled.div`
  display: flex;
  align-items: center;

  ${Button} {
    width: 17rem;
  }
`;

// tslint:disable-next-line: variable-name
export const CTATextLink = styled.span`
  ${textRegularStyle}
  color: ${theme.color.text.white};
  margin-left: 1.5rem;

  a {
    color: ${theme.color.text.white};
    text-decoration: underline;
    margin-left: 0.5rem;
  }
`;

// tslint:disable-next-line: variable-name
export const CTACloseButton = styled(Button)`
  position: absolute;
  top: 47%;
  right: 1%;
  background: transparent;
  min-width: auto;
  padding: 2rem;
  border: none;
  z-index: 1;

  &:hover {
    background: transparent;
  }
`;

// tslint:disable-next-line: variable-name
export const CTACloseIcon = styled(Times)`
  width: 1.8rem;
  color: ${theme.color.text.white};
`;

// tslint:disable-next-line: variable-name
export const CTAGraphic = styled((props) => <img src={ctaGraphic} alt='' {...props} />)`
  margin-left: 3rem;
`;

// tslint:disable-next-line: variable-name
export const CTATopLayer = styled((props) => <img src={topLayer} alt='' {...props} />)`
  position: absolute;
  right: -0.5rem;
  bottom: 0;
`;

// tslint:disable-next-line: variable-name
export const CTAMiddleLayer = styled((props) => <img src={middleLayer} alt='' {...props} />)`
  width: 100%;
  min-width: 160rem;
  position: absolute;
  left: 0;
  bottom: 0;
`;

// tslint:disable-next-line: variable-name
export const CTABottomLayer = styled((props) => <img src={bottomLayer} alt='' {...props} />)`
  width: 100%;
  min-width: 160rem;
  position: absolute;
  left: 0;
  bottom: 2rem;
`;

// tslint:disable-next-line: variable-name
export const CTABackground = styled.div`
  position: relative;
  height: 42rem;
  overflow: hidden;
`;
