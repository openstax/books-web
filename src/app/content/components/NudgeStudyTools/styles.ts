import styled, { css } from 'styled-components/macro';
import { Times } from 'styled-icons/fa-solid/Times/Times';
import { PlainButton } from '../../../components/Button';
import htmlMessage from '../../../components/htmlMessage';
import theme from '../../../theme';
import { remsToPx } from '../../../utils';
import { arrowDesktopHeight, arrowMobileHeight, closeButtonMobileMargin } from './constants';

// tslint:disable-next-line: variable-name
export const NudgeWrapper = styled.div`
  display: contents;
`;

// tslint:disable-next-line: variable-name
export const NudgeContentWrapper = styled.div`
  position: fixed;
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  outline: none;
  ${(props: { top: number, right: number }) => `
    top: ${props.top}px;
    right: ${props.right}px;
  `}
  ${theme.breakpoints.mobile(css`
    right: auto;
    width: 100%;
    text-align: center;
    padding: 0 2rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeContent = styled.div`
  position: relative;
`;

// tslint:disable-next-line: variable-name
const NudgeHeadingStyles = styled.h2`
  font-size: 3.6rem;
  line-height: 1.1;
  letter-spacing: -1.4px;
  margin: 0 0 1.7rem 0;
  color: ${theme.color.text.white};
  overflow: hidden;

  strong {
    color: ${theme.color.primary.yellow.base};
    font-weight: 600;
    white-space: nowrap;
  }

  ${theme.breakpoints.mobile(css`
    font-size: 2.4rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeHeading = htmlMessage('i18n:nudge:study-tools:heading', NudgeHeadingStyles);

// tslint:disable-next-line: variable-name
export const NudgeTextStyles = styled.div`
  font-size: 2.4rem;
  font-weight: 400;
  line-height: 1.1;
  letter-spacing: -1px;
  color: ${theme.color.text.white};
  max-width: 690px;
  overflow: hidden;
  ${theme.breakpoints.mobile(css`
    max-width: 100%;
    font-size: 1.6rem;
    line-height: 1.4;
    padding: 0 2rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeArrow = styled.img`
  position: fixed;
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  display: block;
  height: ${arrowDesktopHeight}rem;
  ${(props: { top: number, left: number }) => `
    top: ${props.top}px;
    left: ${props.left}px;
  `}
  ${theme.breakpoints.mobile(css`
    height: ${arrowMobileHeight}rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeCloseIcon = styled(Times)`
  width: 1.1rem;
  color: ${theme.color.text.white};
`;

// tslint:disable-next-line: variable-name
export const NudgeCloseButton = styled(PlainButton)`
  position: fixed;
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  ${(props: { top: number, left: number }) => `
    top: ${props.top}px;
    left: ${props.left}px;
  `}
  width: 4rem;
  height: 4rem;
  padding: 1rem;
  border-radius: 50%;
  border: 1px solid ${theme.color.white};
  display: flex;
  align-items: center;
  justify-content: center;

  ${theme.breakpoints.mobile(css`
    left: auto;
    top: ${remsToPx(closeButtonMobileMargin)}px;
    right: ${remsToPx(closeButtonMobileMargin)}px;
  `)}
`;

interface NudgeSpotlightPlacement {
  top: number;
  left: number;
  height: number;
  width: number;
}

// tslint:disable-next-line: variable-name
export const NudgeBackground = styled.div`
  position: fixed;
  top: 0;
  width: 100%;
  height: 100%;
  z-index: ${theme.zIndex.nudgeOverlay};
  pointer-events: none;

  display: grid;
  ${(props: NudgeSpotlightPlacement) => `
    grid-template: ${props.top}px ${props.height}px 1fr / ${props.left}px ${props.width}px 1fr;
    grid-template-areas:
      'top       top       top'
      'left   spotlight   right'
      'bottom   bottom   bottom';
  `}
`;

interface ClickBlockerProps {
  area: 'top' | 'right' | 'bottom' | 'left';
}

// tslint:disable-next-line: variable-name
export const ClickBlocker = styled.div`
  width: 100%;
  height: 100%;
  pointer-events: all;
  background-color: ${theme.color.black};
  opacity: 0.9;

  grid-area: ${({area}: ClickBlockerProps) => area};
`;
