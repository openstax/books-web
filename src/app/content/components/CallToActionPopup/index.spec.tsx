import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../test/createTestStore';
import { reactAndFriends } from '../../../../test/utils';
import MessageProvider from '../../../MessageProvider';
import { Store } from '../../../types';
import { assertWindow } from '../../../utils';
import { closeCallToActionPopup, openCallToActionPopup } from '../../actions';
import CallToActionPopup from './';
import { CTACloseButton } from './styles';

describe('CallToActionPopup', () => {
  let store: Store;
  let dispatch: jest.SpyInstance;
  let renderToDom: ReturnType<typeof reactAndFriends>['renderToDom'];

  beforeEach(() => {
    store = createTestStore();

    dispatch = jest.spyOn(store, 'dispatch');
    ({renderToDom} = reactAndFriends());
  });

  it('matches snapshot when close', () => {
    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <CallToActionPopup/>
      </MessageProvider>
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when open', () => {
    store.dispatch(openCallToActionPopup());

    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <CallToActionPopup/>
      </MessageProvider>
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('closes on clicking close button', () => {
    store.dispatch(openCallToActionPopup());

    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <CallToActionPopup/>
      </MessageProvider>
    </Provider>);

    renderer.act(() => {
      const closeButton = component.root.findByType(CTACloseButton);
      closeButton.props.onClick();
    });

    expect(dispatch).toHaveBeenCalledWith(closeCallToActionPopup());
  });

  it('doesnt prevent default on clicking links', () => {
    store.dispatch(openCallToActionPopup());
    const window = assertWindow();
    const {root} = renderToDom(<Provider store={store}>
      <MessageProvider>
        <CallToActionPopup/>
      </MessageProvider>
    </Provider>);

    const loginButton = root.querySelector('[data-testid="log-in"]');
    const signupButton = root.querySelector('[data-testid="sign-up"]');

    expect(loginButton).toBeTruthy();
    expect(signupButton).toBeTruthy();

    const event = window.document.createEvent('MouseEvents');
    event.initEvent('click', true, true);
    const preventDefault = event.preventDefault = jest.fn();

    loginButton.dispatchEvent(event); // this checks for bindings using addEventListener
    ReactTestUtils.Simulate.click(loginButton, {preventDefault}); // this checks for react onClick prop
    signupButton.dispatchEvent(event);
    ReactTestUtils.Simulate.click(signupButton, {preventDefault});

    expect(event.preventDefault).not.toHaveBeenCalled();
  });
});
