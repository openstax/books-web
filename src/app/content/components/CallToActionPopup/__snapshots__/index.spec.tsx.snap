// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallToActionPopup matches snapshot when close 1`] = `null`;

exports[`CallToActionPopup matches snapshot when open 1`] = `
.c15 {
  display: inline-block;
  vertical-align: -.125em;
  overflow: hidden;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  font-size: 1.6rem;
  height: 5rem;
  padding: 0 3rem;
  color: #fff;
  background-color: #f36b32;
  font-weight: bold;
  border: none;
}

.c10:hover {
  background-color: #e96128;
}

.c10:active {
  background-color: #df571e;
}

.c0 {
  height: 34.8rem;
  position: fixed;
  pointer-events: none;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 100;
  -webkit-animation: 800ms bNiYjl ease-out;
  animation: 800ms bNiYjl ease-out;
}

.c4 {
  position: absolute;
  bottom: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 1;
}

.c5 {
  pointer-events: auto;
}

.c6 {
  color: #424242;
  font-size: 3.6rem;
  line-height: 4rem;
  -webkit-letter-spacing: -0.14rem;
  -moz-letter-spacing: -0.14rem;
  -ms-letter-spacing: -0.14rem;
  letter-spacing: -0.14rem;
  color: #fff;
  margin: 0;
  max-width: 50rem;
  overflow: hidden;
}

.c7 {
  color: #424242;
  font-size: 1.8rem;
  line-height: 3rem;
  margin: 1rem 0;
  color: #fff;
  max-width: 57rem;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 .c9 {
  width: 17rem;
}

.c11 {
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  color: #fff;
  margin-left: 1.5rem;
}

.c11 a {
  color: #fff;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  margin-left: 0.5rem;
}

.c13 {
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  background: none;
  pointer-events: auto;
  position: absolute;
  top: 11rem;
  right: 1rem;
  min-width: auto;
  padding: 2rem;
  z-index: 1;
}

.c14 {
  width: 1.8rem;
  color: #fff;
}

.c12 {
  pointer-events: auto;
  margin-left: 3rem;
}

.c3 {
  position: absolute;
  height: 7.7rem;
  width: 48.3rem;
  right: 0;
  bottom: 0;
}

.c2 {
  position: absolute;
  height: 32.8rem;
  width: 100%;
  right: 0;
  bottom: 0;
  left: 0;
}

.c1 {
  position: absolute;
  height: 32.8rem;
  width: 100%;
  top: 0;
  left: 0;
  right: 0;
}

@media screen and (max-width:75em) {
  .c0 {
    display: none;
  }
}

@media print {
  .c0 {
    display: none;
  }
}

<div
  className="c0"
  data-analytics-region="signup CTA"
>
  <img
    alt=""
    className="c1"
    src="bottom-layer.svg"
  />
  <img
    alt=""
    className="c2"
    src="middle-layer.svg"
  />
  <img
    alt=""
    className="c3"
    src="top-layer.svg"
  />
  <div
    className="c4"
  >
    <div
      className="c5"
    >
      <h2
        className="c6"
      >
        Sign up and continue reading. It's 100% free.
      </h2>
      <p
        className="c7"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Read, search, highlight, add notes, build study guides -- all for <strong>free</strong>.",
          }
        }
      />
      <div
        className="c8"
      >
        <a
          className="c9 c10"
          data-analytics-label="sign-up"
          data-testid="sign-up"
          href="/accounts/signup?r=/"
        >
          Sign up
        </a>
        <span
          className="c11"
        >
          or
          <a
            data-analytics-label="login"
            data-testid="login"
            href="/accounts/login?r=/"
          >
            Log in
          </a>
        </span>
      </div>
    </div>
    <img
      alt=""
      className="c12"
      src="desktop-mobile-graphic.svg"
    />
  </div>
  <button
    className="c13"
    data-analytics-label="close"
    data-testid="closeCTA"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="c14 c15"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 352 512"
    >
      <path
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
